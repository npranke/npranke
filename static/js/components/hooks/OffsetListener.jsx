import React from 'react'

function useOffsetListener(elementRef) {
    function reducer() {
        if (elementRef.current) {
            return {
                height: elementRef.current.offsetHeight,
                width: elementRef.current.offsetWidth,
            }
        }

        return {
            height: 0,
            width: 0,
        }
    }

    const [state, dispatch] = React.useReducer(
        reducer,
        {
            height: 0,
            width: 0,
        },
    )

    React.useLayoutEffect(() => {
        function updateOffset() { dispatch() }

        updateOffset()

        window.addEventListener('resize', updateOffset)

        return () => {
            window.removeEventListener('resize', updateOffset)
        }
    }, [])

    return state
}

export default useOffsetListener
