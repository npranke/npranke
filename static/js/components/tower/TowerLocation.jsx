import { DropTarget } from 'react-dnd'
import PropTypes from 'prop-types'
import React from 'react'

import TowerDisk from '@components/tower/TowerDisk'

import useOffsetListener from '@components/hooks/OffsetListener'

export const spec = {
    drop: (props) => {
        return { toLocation: props.location }
    },
    canDrop: (props, monitor) => {
        const diskid = parseInt(monitor.getItem().diskid, 10)
        const topDisk = parseInt(props.diskids[props.diskids.length - 1], 10)

        return (
            !props.isComplete
            && (props.diskids.length === 0 || diskid < topDisk)
        )
    },
}

export function collect(connect) {
    return {
        connectDropTarget: connect.dropTarget(),
    }
}

export function TowerLocation(props) {
    const locationRef = React.useRef(null)
    const { height, width } = useOffsetListener(locationRef)

    let diskHeight, diskWidth, scaleFactor

    if (props.isPortrait) {
        diskHeight = 904
        diskWidth = 329
        scaleFactor = 1

        if (diskHeight > height) {
            scaleFactor = height / diskHeight
            diskWidth *= scaleFactor
            diskHeight = height
        }

        const availableWidth = width * 0.7
        const disksWidth = diskWidth * props.disks

        if (disksWidth > availableWidth) {
            scaleFactor = availableWidth / disksWidth
            diskWidth = availableWidth / props.disks
            diskHeight *= scaleFactor
        }
    } else {
        diskHeight = 329
        diskWidth = 904
        scaleFactor = 1

        if (diskWidth > width) {
            scaleFactor = width / diskWidth
            diskHeight *= scaleFactor
            diskWidth = width
        }

        const availableHeight = height * 0.7
        const disksHeight = diskHeight * props.disks

        if (disksHeight > availableHeight) {
            scaleFactor = availableHeight / disksHeight
            diskHeight = availableHeight / props.disks
            diskWidth *= scaleFactor
        }
    }

    let towerDisks = props.tower[
        props.isPortrait ? 'portrait' : 'landscape'
    ].filter((disk) => {
        return props.diskids.indexOf(disk.diskid) !== -1
    }).map((disk) => {
        return (
            <React.Fragment key={ disk.diskid }>
                <div
                    className="disk-wrapper"
                    style={ { height: diskHeight, width: diskWidth } }
                >
                    <TowerDisk
                        diskid={ disk.diskid }
                        diskids={ props.diskids }
                        image={ disk.image }
                        isComplete={ props.isComplete }
                        moveDisk={ props.moveDisk }
                        location={ props.location }
                    />
                </div>
            </React.Fragment>
        )
    })

    if (props.isPortrait) {
        towerDisks.reverse()
    }

    if (props.isComplete) {
        towerDisks = (
            <a
                className="location-section"
                onClick={ props.tower.clickHandler }
                href={ props.tower.link }
                target="_blank"
                rel="noreferrer noopener external"
                aria-label="NASA/JPL Space Images"
            >
                { towerDisks }
            </a>
        )
    }

    towerDisks = (
        <div
            className={
                'location-section disks '
                + `${props.location} `
                + `${props.isComplete ? 'complete' : 'notcomplete'}`
            }
        >
            { towerDisks }
        </div>
    )

    return props.isPortrait ? props.connectDropTarget(
        <div className="location" ref={ locationRef }>
            <div className="location-section name">
                <div className="name-text">
                    { props.location }
                </div>
            </div>
            { towerDisks }
        </div>,
    ) : props.connectDropTarget(
        <div className="location" ref={ locationRef }>
            { towerDisks }
            <div className="location-section name">
                <div className="name-text">
                    { props.location }
                </div>
            </div>
        </div>,
    )
}

TowerLocation.defaultProps = {
    connectDropTarget: undefined,
    diskids: [],
    isComplete: false,
    isPortrait: false,
}

TowerLocation.propTypes = {
    connectDropTarget: PropTypes.func,
    diskids: PropTypes.arrayOf(PropTypes.string),
    disks: PropTypes.number.isRequired,
    isComplete: PropTypes.bool,
    isPortrait: PropTypes.bool,
    location: PropTypes.string.isRequired,
    moveDisk: PropTypes.func.isRequired,
    tower: PropTypes.shape({
        clickHandler: PropTypes.func.isRequired,
        landscape: PropTypes.arrayOf(
            PropTypes.shape({
                diskid: PropTypes.string.isRequired,
                image: PropTypes.node.isRequired,
            }).isRequired,
        ).isRequired,
        link: PropTypes.string.isRequired,
        portrait: PropTypes.arrayOf(
            PropTypes.shape({
                diskid: PropTypes.string.isRequired,
                image: PropTypes.node.isRequired,
            }).isRequired,
        ).isRequired,
    }).isRequired,
}

export default DropTarget('disk', spec, collect)(TowerLocation)
