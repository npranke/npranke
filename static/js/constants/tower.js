import HTML5Backend from 'react-dnd-html5-backend'
import { MouseTransition, TouchTransition } from 'react-dnd-multi-backend'
import TouchBackend from 'react-dnd-touch-backend'

import EmptyDiscPreview from '@img/tower-empty-disc-preview.gif'

import Landscape1of2 from '@img/tower-landscape-22931-two-1-of-2.jpg'
import Landscape2of2 from '@img/tower-landscape-22931-two-2-of-2.jpg'
import Portrait1of2 from '@img/tower-portrait-22931-two-1-of-2.jpg'
import Portrait2of2 from '@img/tower-portrait-22931-two-2-of-2.jpg'

import Landscape1of3 from '@img/tower-landscape-22938-three-1-of-3.jpg'
import Landscape2of3 from '@img/tower-landscape-22938-three-2-of-3.jpg'
import Landscape3of3 from '@img/tower-landscape-22938-three-3-of-3.jpg'
import Portrait1of3 from '@img/tower-portrait-22938-three-1-of-3.jpg'
import Portrait2of3 from '@img/tower-portrait-22938-three-2-of-3.jpg'
import Portrait3of3 from '@img/tower-portrait-22938-three-3-of-3.jpg'

import Landscape1of4 from '@img/tower-landscape-22944-four-1-of-4.jpg'
import Landscape2of4 from '@img/tower-landscape-22944-four-2-of-4.jpg'
import Landscape3of4 from '@img/tower-landscape-22944-four-3-of-4.jpg'
import Landscape4of4 from '@img/tower-landscape-22944-four-4-of-4.jpg'
import Portrait1of4 from '@img/tower-portrait-22944-four-1-of-4.jpg'
import Portrait2of4 from '@img/tower-portrait-22944-four-2-of-4.jpg'
import Portrait3of4 from '@img/tower-portrait-22944-four-3-of-4.jpg'
import Portrait4of4 from '@img/tower-portrait-22944-four-4-of-4.jpg'

import Landscape1of5 from '@img/tower-landscape-22423-five-1-of-5.jpg'
import Landscape2of5 from '@img/tower-landscape-22423-five-2-of-5.jpg'
import Landscape3of5 from '@img/tower-landscape-22423-five-3-of-5.jpg'
import Landscape4of5 from '@img/tower-landscape-22423-five-4-of-5.jpg'
import Landscape5of5 from '@img/tower-landscape-22423-five-5-of-5.jpg'
import Portrait1of5 from '@img/tower-portrait-22423-five-1-of-5.jpg'
import Portrait2of5 from '@img/tower-portrait-22423-five-2-of-5.jpg'
import Portrait3of5 from '@img/tower-portrait-22423-five-3-of-5.jpg'
import Portrait4of5 from '@img/tower-portrait-22423-five-4-of-5.jpg'
import Portrait5of5 from '@img/tower-portrait-22423-five-5-of-5.jpg'

export const DRAG_AND_DROP_BACKEND_PIPELINE = {
    backends: [
        {
            backend: HTML5Backend,
        },
        {
            backend: TouchBackend,
            preview: false,
            transition: TouchTransition,
        },
        {
            backend: HTML5Backend,
            preview: false,
            transition: MouseTransition,
        },
    ],
}

export const EMPTY_PREVIEW_SOURCE = EmptyDiscPreview

export const LAYOUTS = ['landscape', 'portrait']

export const LOCATIONS = ['origin', 'buffer', 'target']

export const TOWER_SOURCES = {
    2: {
        disks: {
            landscape: [
                Landscape1of2,
                Landscape2of2,
            ],
            portrait: [
                Portrait1of2,
                Portrait2of2,
            ],
        },
        sourceid: 22931,
    },
    3: {
        disks: {
            landscape: [
                Landscape1of3,
                Landscape2of3,
                Landscape3of3,
            ],
            portrait: [
                Portrait1of3,
                Portrait2of3,
                Portrait3of3,
            ],
        },
        sourceid: 22938,
    },
    4: {
        disks: {
            landscape: [
                Landscape1of4,
                Landscape2of4,
                Landscape3of4,
                Landscape4of4,
            ],
            portrait: [
                Portrait1of4,
                Portrait2of4,
                Portrait3of4,
                Portrait4of4,
            ],
        },
        sourceid: 22944,
    },
    5: {
        disks: {
            landscape: [
                Landscape1of5,
                Landscape2of5,
                Landscape3of5,
                Landscape4of5,
                Landscape5of5,
            ],
            portrait: [
                Portrait1of5,
                Portrait2of5,
                Portrait3of5,
                Portrait4of5,
                Portrait5of5,
            ],
        },
        sourceid: 22423,
    },
}
