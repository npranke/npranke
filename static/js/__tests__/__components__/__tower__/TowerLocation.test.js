import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'

import { collect, spec, TowerLocation } from '@components/tower/TowerLocation'

import * as OffsetListener from '@components/hooks/OffsetListener'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('@components/tower/TowerDisk', () => {
    return 'TowerDisk'
})

const towerLayouts = ['landscape', 'portrait']
const towers = Array.from(
    { length: 4 },
    (value, index) => { return index + 2 },
).reduce((towerAccumulator, tower) => {
    towerAccumulator[tower] = {
        clickHandler: jest.fn(),
        link: 'tower-link',
    }

    towerLayouts.forEach((layout) => {
        towerAccumulator[tower][layout] = Array.from(
            { length: tower },
            (value, index) => {
                return {
                    diskid: `${index + 1}`,
                    image: (
                        <img src="" alt="alt-text" className="disk-image" />
                    ),
                }
            },
        )
    })

    return towerAccumulator
}, {})

describe('TowerLocation', () => {
    describe('drop target collect', () => {
        let connect

        beforeAll(() => {
            connect = {
                dropTarget: jest.fn(),
            }
        })

        afterEach(() => {
            connect.dropTarget.mockReset()
        })

        test('collect is a function', () => {
            expect(collect).toBeInstanceOf(Function)
        })

        test('returns object with expected key', () => {
            const collectResult = collect(connect)
            const collectResultKeys = Object.keys(collectResult)

            expect(collectResult).toBeInstanceOf(Object)
            expect(
                collectResultKeys.length,
            ).toBe(1)
            expect(
                collectResultKeys.indexOf('connectDropTarget'),
            ).toBeGreaterThan(-1)
        })

        test('calls connect.dropTarget', () => {
            collect(connect)

            expect(
                connect.dropTarget,
            ).toHaveBeenCalledTimes(1)
        })
    })

    describe('drop target spec', () => {
        let props, monitor

        beforeAll(() => {
            monitor = {
                getItem: jest.fn(),
            }
        })

        beforeEach(() => {
            props = {
                diskids: ['3', '2', '1'],
                disks: 3,
                isComplete: false,
                location: 'origin',
                moveDisk: jest.fn(),
                tower: towers[3],
            }
        })

        afterEach(() => {
            props.moveDisk.mockReset()
            props.tower.clickHandler.mockReset()
            monitor.getItem.mockReset()
        })

        test('spec is an object with expected 2 keys', () => {
            const specKeys = Object.keys(spec)

            expect(spec).toBeInstanceOf(Object)
            expect(
                Object.keys(spec).length,
            ).toBe(2)
            expect(
                specKeys.indexOf('drop'),
            ).toBeGreaterThan(-1)
            expect(
                specKeys.indexOf('canDrop'),
            ).toBeGreaterThan(-1)
        })

        test('drop returns expected object', () => {
            props.location = 'buffer'

            const dropResult = spec.drop(props)

            expect(
                dropResult,
            ).toEqual({ toLocation: 'buffer' })
        })

        test('canDrop returns false when isComplete true', () => {
            monitor.getItem.mockImplementationOnce(
                () => { return { diskid: '1' } },
            )
            props.diskids = ['3', '2', '1']
            props.isComplete = true
            props.location = 'target'

            const canDropResult = spec.canDrop(props, monitor)

            expect(canDropResult).toBe(false)
        })

        test('canDrop returns false when disk is larger than top', () => {
            monitor.getItem.mockImplementationOnce(
                () => { return { diskid: '2' } },
            )
            props.diskids = ['1']
            props.isComplete = false
            props.location = 'target'

            const canDropResult = spec.canDrop(props, monitor)

            expect(canDropResult).toBe(false)
        })

        test('canDrop returns true when disk is smaller than top', () => {
            monitor.getItem.mockImplementationOnce(
                () => { return { diskid: '2' } },
            )
            props.diskids = ['3']
            props.isComplete = false
            props.location = 'target'

            const canDropResult = spec.canDrop(props, monitor)

            expect(canDropResult).toBe(true)
        })

        test('canDrop returns true when diskids is empty', () => {
            monitor.getItem.mockImplementationOnce(
                () => { return { diskid: '3' } },
            )
            props.diskids = []
            props.isComplete = false
            props.location = 'target'

            const canDropResult = spec.canDrop(props, monitor)

            expect(canDropResult).toBe(true)
        })
    })

    test('has one location', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                disks={ 3 }
                location="buffer"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.location').length,
        ).toBe(1)
    })

    test('has two location sections', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                disks={ 3 }
                location="buffer"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.location-section').length,
        ).toBe(2)
    })

    test('has name section with name text', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                disks={ 3 }
                location="buffer"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.location-section.name').length,
        ).toBe(1)
        expect(
            towerLocation.find('.location-section.name .name-text').text(),
        ).toEqual('buffer')
    })

    test('has disks section with specific location class', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                disks={ 3 }
                location="buffer"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.location-section.disks').length,
        ).toBe(1)
        expect(
            towerLocation.find('.location-section.disks.buffer').length,
        ).toBe(1)
    })

    test('disks section has complete class when isComplete', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['3', '2', '1'] }
                disks={ 3 }
                isComplete
                location="target"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disks.target.complete').length,
        ).toBe(1)
    })

    test('disks section has notcomplete class when not isComplete', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['3', '2', '1'] }
                disks={ 3 }
                location="origin"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disks.origin.notcomplete').length,
        ).toBe(1)
    })

    test('disks section has link when isComplete', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['3', '2', '1'] }
                disks={ 3 }
                isComplete
                location="target"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disks a').length,
        ).toBe(1)
        expect(
            towerLocation.find('.disks a').first().props().href,
        ).toEqual('tower-link')
    })

    test('disks section does not have link when not isComplete', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['3', '2', '1'] }
                disks={ 3 }
                location="origin"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disks a').length,
        ).toBe(0)
    })

    test('has no disk wrappers with no disks', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ [] }
                disks={ 3 }
                location="origin"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disk-wrapper TowerDisk').length,
        ).toBe(0)
    })

    test('has disk wrappers with some disks', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['2', '1'] }
                disks={ 3 }
                location="buffer"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disk-wrapper TowerDisk').length,
        ).toBe(2)
    })

    test('has disk wrappers with all disks', () => {
        const towerLocation = shallow(
            <TowerLocation
                connectDropTarget={
                    jest.fn((locationNode) => { return locationNode })
                }
                diskids={ ['3', '2', '1'] }
                disks={ 3 }
                isComplete
                location="target"
                moveDisk={ jest.fn() }
                tower={ towers[3] }
            />,
        )

        expect(
            towerLocation.find('.disk-wrapper TowerDisk').length,
        ).toBe(3)
    })

    describe('setting disk height and width', () => {
        let useOffsetListener

        beforeAll(() => {
            useOffsetListener = jest.spyOn(OffsetListener, 'default')
        })

        afterEach(() => {
            useOffsetListener.mockClear()
        })

        afterAll(() => {
            useOffsetListener.mockRestore()
        })

        describe('when not isPortrait', () => {
            test('diskWidth greater than location width', () => {
                useOffsetListener.mockReturnValue(
                    { height: 1645, width: 452 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2'] }
                        disks={ 3 }
                        isComplete={ false }
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 164.5, width: 452 })
                    },
                )
            })

            test('diskWidth not greater than location width', () => {
                useOffsetListener.mockReturnValue(
                    { height: 3290, width: 904 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2', '1'] }
                        disks={ 3 }
                        isComplete
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 329, width: 904 })
                    },
                )
            })

            test('disks height greater than availableHeight', () => {
                useOffsetListener.mockReturnValue(
                    { height: 822.5, width: 904 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3'] }
                        disks={ 3 }
                        isComplete={ false }
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 164.5, width: 452 })
                    },
                )
            })

            test('disks height not greater than availableHeight', () => {
                useOffsetListener.mockReturnValue(
                    { height: 1645, width: 904 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2', '1'] }
                        disks={ 3 }
                        isComplete
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 329, width: 904 })
                    },
                )
            })
        })

        describe('when isPortrait', () => {
            test('diskHeight greater than location height', () => {
                useOffsetListener.mockReturnValue(
                    { height: 452, width: 1645 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2'] }
                        disks={ 3 }
                        isComplete={ false }
                        isPortrait
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 452, width: 164.5 })
                    },
                )
            })

            test('diskHeight not greater than location height', () => {
                useOffsetListener.mockReturnValue(
                    { height: 904, width: 3290 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2', '1'] }
                        disks={ 3 }
                        isComplete
                        isPortrait
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 904, width: 329 })
                    },
                )
            })

            test('disks width greater than availableWidth', () => {
                useOffsetListener.mockReturnValue(
                    { height: 904, width: 822.5 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3'] }
                        disks={ 3 }
                        isComplete={ false }
                        isPortrait
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 452, width: 164.5 })
                    },
                )
            })

            test('disks width not greater than availableWidth', () => {
                useOffsetListener.mockReturnValue(
                    { height: 904, width: 1645 },
                )

                const towerLocation = shallow(
                    <TowerLocation
                        connectDropTarget={
                            jest.fn((locationNode) => { return locationNode })
                        }
                        diskids={ ['3', '2', '1'] }
                        disks={ 3 }
                        isComplete
                        isPortrait
                        location="target"
                        moveDisk={ jest.fn() }
                        tower={ towers[3] }
                    />,
                )

                towerLocation.find('.disk-wrapper').forEach(
                    (diskWrapper) => {
                        expect(
                            diskWrapper.props().style,
                        ).toEqual({ height: 904, width: 329 })
                    },
                )
            })
        })
    })
})

describe('TowerLocation snapshot', () => {
    describe('when not isPortrait', () => {
        test('matches snapshot when empty', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ [] }
                    disks={ 2 }
                    location="buffer"
                    moveDisk={ jest.fn() }
                    tower={ towers[2] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with some disks', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['1'] }
                    disks={ 2 }
                    location="buffer"
                    moveDisk={ jest.fn() }
                    tower={ towers[2] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with all disks and not target', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['2', '1'] }
                    disks={ 2 }
                    isComplete={ false }
                    location="origin"
                    moveDisk={ jest.fn() }
                    tower={ towers[2] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with all disks and target', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['2', '1'] }
                    disks={ 2 }
                    isComplete
                    location="target"
                    moveDisk={ jest.fn() }
                    tower={ towers[2] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })
    })

    describe('when isPortrait', () => {
        test('matches snapshot when empty', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ [] }
                    disks={ 4 }
                    isPortrait
                    location="buffer"
                    moveDisk={ jest.fn() }
                    tower={ towers[4] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with some disks', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['3', '2'] }
                    disks={ 4 }
                    isPortrait
                    location="buffer"
                    moveDisk={ jest.fn() }
                    tower={ towers[4] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with all disks and not target', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['4', '3', '2', '1'] }
                    disks={ 4 }
                    isComplete={ false }
                    isPortrait
                    location="origin"
                    moveDisk={ jest.fn() }
                    tower={ towers[4] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })

        test('matches snapshot with all disks and target', () => {
            const towerLocation = renderer.create(
                <TowerLocation
                    connectDropTarget={
                        jest.fn((locationNode) => { return locationNode })
                    }
                    diskids={ ['4', '3', '2', '1'] }
                    disks={ 4 }
                    isComplete
                    isPortrait
                    location="target"
                    moveDisk={ jest.fn() }
                    tower={ towers[4] }
                />,
            ).toJSON()

            expect(towerLocation).toMatchSnapshot()
        })
    })
})
