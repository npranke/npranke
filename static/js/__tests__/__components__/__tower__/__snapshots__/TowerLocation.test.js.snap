// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TowerLocation snapshot when isPortrait matches snapshot when empty 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        buffer
      </div>
    </div>
    <div
      class="location-section disks buffer notcomplete"
    />
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when isPortrait matches snapshot with all disks and not target 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        origin
      </div>
    </div>
    <div
      class="location-section disks origin notcomplete"
    >
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when isPortrait matches snapshot with all disks and target 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        target
      </div>
    </div>
    <div
      class="location-section disks target complete"
    >
      <a
        aria-label="NASA/JPL Space Images"
        class="location-section"
        href="tower-link"
        rel="noreferrer noopener external"
        target="_blank"
      >
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when isPortrait matches snapshot with some disks 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        buffer
      </div>
    </div>
    <div
      class="location-section disks buffer notcomplete"
    >
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when not isPortrait matches snapshot when empty 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section disks buffer notcomplete"
    />
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        buffer
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when not isPortrait matches snapshot with all disks and not target 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section disks origin notcomplete"
    >
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
    </div>
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        origin
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when not isPortrait matches snapshot with all disks and target 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section disks target complete"
    >
      <a
        aria-label="NASA/JPL Space Images"
        class="location-section"
        href="tower-link"
        rel="noreferrer noopener external"
        target="_blank"
      >
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
        <div
          class="disk-wrapper"
          style="height: 0px; width: 0px;"
        >
          TowerDisk
        </div>
      </a>
    </div>
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        target
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TowerLocation snapshot when not isPortrait matches snapshot with some disks 1`] = `
<DocumentFragment>
  <div
    class="location"
  >
    <div
      class="location-section disks buffer notcomplete"
    >
      <div
        class="disk-wrapper"
        style="height: 0px; width: 0px;"
      >
        TowerDisk
      </div>
    </div>
    <div
      class="location-section name"
    >
      <div
        class="name-text"
      >
        buffer
      </div>
    </div>
  </div>
</DocumentFragment>
`;
